src/cpp/src/llm_pipeline.cpp

```diff
@@ -244,60 +233,84 @@ class StatefulLLMPipeline final : public LLMPipelineImplBase {
+        bool enable_prefix_caching = false;
 
-                sequence_group->set_sequence_group_ptr(sequence_group);
-                requests.push_back(sequence_group);
+        for (size_t request_id = 0; request_id < batch_size; request_id++) {
+            SequenceGroup::Ptr sequence_group;
+            if (is_chat_conversation) {
+                sequence_group = std::make_shared<SequenceGroup>(request_id, m_tokenized_chat_history, config, block_size, enable_prefix_caching);
+                sequence_group->update_processed_tokens_num(m_tokenized_chat_history.get_shape().at(1) - 1);
+            } else {
+                size_t seq_len = input_ids.get_shape().at(1);
+                size_t batch_offset = request_id * seq_len;
+                const int64_t* prompt_start = input_ids.data<const int64_t>() + batch_offset;
+                std::vector<int64_t> tokenized_prompt(prompt_start, prompt_start + seq_len);
+                // in case of multi batch scenario, remove eos_token_id at start of prompt
+                auto real_prompt_start = std::find_if(tokenized_prompt.begin(), tokenized_prompt.end(), [&config](int64_t token) { return token != config.eos_token_id; });
+                tokenized_prompt.erase(tokenized_prompt.begin(), real_prompt_start);
+
+                sequence_group = std::make_shared<SequenceGroup>(request_id, tokenized_prompt, config, block_size, enable_prefix_caching);
+                sequence_group->update_processed_tokens_num(tokenized_prompt.size() - 1);
             }
 
-            Sampler sampler = Sampler(m_tokenizer);
-            std::tie(result, m_selected_beam) = ov::genai::get_lm_encoded_results(m_model_runner, input_ids, concatenated_attention_mask, streamer_ptr,
-                                                                                  sampler, requests, position_ids, std::nullopt, m_selected_beam);
+            sequence_group->set_sequence_group_ptr(sequence_group);
+            requests.push_back(sequence_group);
         }
 
-        if (!is_chat_conversation) {
+        Sampler sampler = Sampler(m_tokenizer);
+        // we can't properly refer to history in case of chat scenario with beam search, so reset_kv_state and use the whole history for each new propmt
+        auto input_tokens = (is_chat_conversation && config.is_beam_search()) ? m_tokenized_chat_history : input_ids;
+        result = ov::genai::get_lm_encoded_results(m_model_runner, input_tokens, concatenated_attention_mask, streamer_ptr, sampler, requests, position_ids, std::nullopt);
+
+        if (!is_chat_conversation || config.is_beam_search()) {
             reset_kv_state();
-            m_selected_beam = std::nullopt;
-        } else {
+        }
+
+        if (is_chat_conversation) {
             m_is_cache_empty = false;
+
+            // remove eos token, if it is at the end
+            auto last_token = result.tokens[0].back() == config.eos_token_id ? result.tokens[0].size() - 1 : result.tokens[0].size();
```

Comment:
1. why do we need to drop EOS at all?
2. we can have multiple EOS tokens (stop_token_ids in generation config)

Reply:
when we call `generation` with `StringInputs` in chat mode with cache, it always will be added eos token to sequence to the beginning https://github.com/openvinotoolkit/openvino.genai/blob/master/src/cpp/src/llm_pipeline.cpp#L114

if `generation` was finished on the prev step because it faced with `eos token` -> we will have two `eos token` in the history , if next iteration go through `StringInputs` 
if `generation` was finished because of `max_token` or any other stop token, `eos token` will be added and it will work 

I'm not sure how to handle it correctly, maybe it is batter to check if we have double eos, when we put propmt to history ?

Reply:
> if generation was finished because of max_token or any other stop token, eos token will be added and it will work

Does that mean that `eos_token_id` and `stop_token_ids` behave differently?